akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ShardingState.scala

```diff
@@ -3,27 +3,58 @@
  */
 
 package akka.cluster.sharding.typed.internal
+
 import akka.actor.typed.Behavior
 import akka.actor.typed.scaladsl.Behaviors
-import akka.cluster.sharding.{ ClusterSharding, ShardRegion }
-import akka.cluster.sharding.ShardRegion.CurrentShardRegionState
 import akka.actor.typed.scaladsl.adapter._
 import akka.annotation.InternalApi
-import akka.cluster.sharding.typed.{ ClusterShardingQuery, GetShardRegionState }
+import akka.cluster.sharding.ShardRegion.CurrentShardRegionState
+import akka.cluster.sharding.typed.{ ClusterShardingQuery, GetClusterShardingStats, GetShardRegionState }
+import akka.cluster.sharding.{ ClusterSharding, ShardRegion }
+import akka.pattern.AskTimeoutException
+import akka.util.Timeout
+
+import scala.concurrent.Future
 
 /**
  * INTERNAL API
  */
 @InternalApi private[akka] object ShardingState {
 
-  def behavior(classicSharding: ClusterSharding): Behavior[ClusterShardingQuery] = Behaviors.receiveMessage {
-    case GetShardRegionState(key, replyTo) =>
-      if (classicSharding.getShardTypeNames.contains(key.name)) {
-        classicSharding.shardRegion(key.name).tell(ShardRegion.GetShardRegionState, replyTo.toClassic)
-      } else {
-        replyTo ! CurrentShardRegionState(Set.empty)
+  def behavior(classicSharding: ClusterSharding): Behavior[ClusterShardingQuery] =
+    Behaviors.setup { context =>
+      Behaviors.receiveMessage {
+        case GetShardRegionState(key, replyTo) =>
+          if (classicSharding.getShardTypeNames.contains(key.name)) {
+            classicSharding.shardRegion(key.name).tell(ShardRegion.GetShardRegionState, replyTo.toClassic)
+          } else {
+            replyTo ! CurrentShardRegionState(Set.empty)
+          }
+          Behaviors.same
+
+        case GetClusterShardingStats(timeout, replyTo) =>
+          import akka.pattern.{ ask, pipe }
+          implicit val t: Timeout = timeout
+          implicit val ec = context.system.executionContext
+
+          val regions = classicSharding.shardTypeNames.map(classicSharding.shardRegion)
+
+          if (regions.nonEmpty) {
+            Future
+              .firstCompletedOf(regions.map { region =>
+                (region ? ShardRegion.GetClusterShardingStats(timeout)).mapTo[ShardRegion.ClusterShardingStats]
```

Comment:
Because we only need one region I believe, to reach the `ShardCoordinator` and run the query for cluster stats.

Reply:
I think it's wrong to query some arbitrary region. Each region here corresponds to one entity type, and the GetClusterShardingStats should be a query for a specific entity type. GetClusterShardingStats should include the entityTypeKey in same way as GetShardRegionState. Then the entityTypeKey is used for selecting corresponding region (in same way as for GetShardRegionState).